pkg/scheduler/framework/runtime/framework.go

```diff
@@ -346,12 +330,39 @@ func NewFramework(r Registry, profile *config.KubeSchedulerProfile, opts ...Opti
 		fillEventToPluginMap(p, options.clusterEventMap)
 	}
 
+	// initialize plugins per individual extension points
 	for _, e := range f.getExtensionPoints(profile.Plugins) {
 		if err := updatePluginList(e.slicePtr, *e.plugins, pluginsMap); err != nil {
 			return nil, err
 		}
 	}
 
+	// initialize multiPoint plugins to their expanded extension points
+	if len(profile.Plugins.MultiPoint.Enabled) > 0 {
```

Comment:
Why add this condition? What if the multiPoint only carries `disabled`? Given the following pluginSet:

```go
			customPlugins: &v1beta3.Plugins{
				MultiPoint: v1beta3.PluginSet{
					Disabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin1"},
						{Name: "DefaultPlugin2"},
					},
				},
			},
			defaultPlugins: &v1beta3.Plugins{
				Filter: v1beta3.PluginSet{
					Enabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin1"},
					},
				},
				Score: v1beta3.PluginSet{
					Enabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin2"},
					},
				},
			},
```

The merged pluginSet would be:

```go
				Filter: v1beta3.PluginSet{
					Enabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin1"},
					},
				},
				Score: v1beta3.PluginSet{
					Enabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin2"},
					},
				},
				MultiPoint: v1beta3.PluginSet{
					Disabled: []v1beta3.Plugin{
						{Name: "DefaultPlugin1"},
						{Name: "DefaultPlugin2"},
					},
				},
```

In this case, we should still handle MultiPoint even if it only has `disabled`, right? Because without it, the result would carry regular extension point **only** - which is processed earlier in updatePluginList().

Reply:
MultiPoint only takes effect if it has Enabled plugins. Otherwise, there will be no MultiPoint expansion. Putting plugins into `multiPoint.Disabled` without `multiPoint.Enabled` doesn't do anything. As you showed in your example, those plugins are specifically enabled through individual extension points.

Reply:
can we resolve this one?

Reply:
Bump. (`mergePluginSet()` in default_plugins.go looks good now)

And make sure to have UTs covering this.

Reply:
This was covered by other tests, the `disabled` field is carried in. I added a new test specifically showing this without any other settings.

Reply:
sg, @Huang-Wei  good to resolve this comment?

Reply:
A corner case here: what if pluginA is enabled AND also disabled in multipoint? In the case below, we should disable scorePlugin1, right? (maybe it's not a _corner_ case, as the plugins here is merged from default plus user specified ones - for example, default plugins enabled scorePlugin1, but user disables it via CC)

```go
				MultiPoint: config.PluginSet{
					Enabled: []config.Plugin{
						{Name: queueSortPlugin},
						{Name: bindPlugin},
						{Name: scorePlugin1},
					},
					Disabled: []config.Plugin{
						{Name: scorePlugin1},
					},
				},
```

Reply:
Isn't that this test? https://github.com/kubernetes/kubernetes/pull/105611/files#diff-392546fd01ec251b85f84db8e534ad5d22aa4e58de40b629417d0af0f90916b9R390-R432

Reply:
> for example, default plugins enabled scorePlugin1, but user disables it via CC)

that's the exact test case I shared. The user disabling it takes precedence as defined by the hierarchy of `Specific > MultiPoint > Default(MultiPoint)`

Reply:
My point is that scorePlugin1 should be disabled in this case.

Reply:
It is disabled, the user disabling it via CC takes precedence. See that in that test, the user disables `DefaultPlugin2` via CC, and the expected output shows `DefaultPlugin2` disabled. This is the same scenario.

Reply:
In other words, in terms of multiPoint, we can discard `disabled: *`, but not for `disabled: <specific plugin>`.

Reply:
No, both of those are valid in certain use cases and there are test cases for both.

Reply:
> No, both of those are valid in certain use cases and there are test cases for both.

Without logic of dealing with `multipoint.disabled`, how can you achieve that? Am I missing something?

Also, the `scorePlugin1` is still there in the wantPlugins, in subtest "disable and enable MultiPoint plugin by name".

Reply:
> Without logic of dealing with multipoint.disabled, how can you achieve that? Am I missing something?

`mergePlugins()` handles this when the component config is processed. If the plugin is disabled, it won't be added to the merged config that gets processed here. There is no need for us to handle `multiPoint.disabled` at this point because it has already been merged.

> Also, the scorePlugin1 is still there in the wantPlugins, in subtest "disable and enable MultiPoint plugin by name".

The difference in that test is that the user *enabled* that plugin as well. Enabling always takes precedence over disabling. This allows for overriding and is consistent with the current behavior.

Reply:
See comments in the godoc: https://github.com/kubernetes/kubernetes/pull/105611/files#diff-5aed18dafd438aa779df20e2474aeeaa6a75ef70493c282b4c428acf90553382R200-R202

Reply:
The reason we only need to act in the framework runtime if `len(multiPoint.enabled) > 0` is because there will be nothing to do if there are no plugins enabled through multiPoint. If there were (either custom or default), they will have been merged into the single config here. We do check some of the disabled plugins in this section to make sure the plugin isn't specifically disabled by name. But if there are no `multiPoint` plugins enabled, then we don't care about `multiPoint.disabled`.

Reply:
`multiPoint.disabled` isn't just shorthand to disable every plugin in every extension point. It is only for disabling the multiPoint behavior, ie the expanding into those extension points

Reply:
> `mergePlugins()` handles this when the component config is processed. If the plugin is disabled, it won't be added to the merged config that gets processed here.

> If there were disabled plugins in multiPoint (either custom or default), they will have been merged into the single config here.

This makes sense to me. Thanks for the explanation.